### PR TITLE
DEV: Create posts from form templates

### PR DESCRIPTION
**Pull request by keegangeorge**
Note: This PR is an addition to the `experimental_form_templates` feature.

In a previous PR (https://github.com/discourse/discourse/pull/21190), we added the ability to show a category's form template form in the composer. In **this PR** we add the logic for processing the form data and turning it into a topic. This PR also adds basic support for validations defined in the templates YAML markup.

https://github.com/discourse/discourse/assets/30090424/173d708d-9fa6-4953-a79e-e3f929e1c5c0

### CODE REVIEW

File: app/assets/javascripts/discourse/app/services/composer.js [hunk 1]
```diff
@@ -1663,4 +1684,92 @@ export default class ComposerController extends Controller {
   clearLastValidatedAt() {
     this.set("lastValidatedAt", null);
   }
+
+  prepareFormTemplateData(event, skipValidation = false) {
```
Comment (pmusaraj) [inline app/assets/javascripts/discourse/app/services/composer.js:1688]: If possible, I'd like to have the validations live outside the composer service to keep things lean in the composer.

**Comment by pmusaraj** [comment]
I am seeing an issue replying to a topic as well: 

image

File: app/assets/javascripts/discourse/app/services/composer.js [hunk 2]
```diff
@@ -1663,4 +1684,92 @@ export default class ComposerController extends Controller {
...
+      const value = item[key];
+      if (value) {
+        return `### ${key}\n${value}`;
+      }
+    });
+
+    this.model.set("reply", formattedOutput.join("\n\n"));
+  }
+
+  _validateFormTemplateData(form) {
```
[the remark this answers is not in the document]
Comment (keegangeorge) [reply app/assets/javascripts/discourse/app/services/composer.js:1736]: I added some system specs for the validations [here](https://github.com/discourse/discourse/blob/150f8862cfe7ef52a7e07a17a07190d339582f4a/spec/system/composer/template_validation_spec.rb). Should the unit tests be added in addition to those?

File: app/assets/javascripts/discourse/app/services/composer.js [hunk 3]
```diff
@@ -788,6 +795,12 @@ export default class ComposerController extends Controller {
 
   @action
   saveAction(ignore, event) {
+    if (this.siteSettings.experimental_form_templates) {
+      if (this.formTemplateIds?.length > 0) {
+        this.prepareFormTemplateData(event);
```
[the remark this answers is not in the document]
Comment (pmusaraj) [reply app/assets/javascripts/discourse/app/services/composer.js:800]: Ah! Thanks, I will test with that.

File: app/assets/javascripts/discourse/app/services/composer.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (pmusaraj) [reply app/assets/javascripts/discourse/app/services/composer.js:1736]: Yes, I think so. System tests are good for one or two "happy paths". Unit tests are good for going through all the cases and making sure the validation returns results as expected. 

Can be in a separate PR, though.

**Review by pmusaraj — APPROVED** [review]
There's still a few minor details to iron out, but validations work locally for me with the correct syntax, approving. Thanks!

File: app/assets/javascripts/discourse/app/components/composer-messages.js [hunk 4]
```diff
@@ -183,6 +183,11 @@ export default class ComposerMessages extends Component {
       return;
     }
 
+    // We don't care about similar topics when creating with a form template
+    if (this.composer?.category?.form_template_ids.length > 0) {
```
Comment (pmusaraj) [inline app/assets/javascripts/discourse/app/components/composer-messages.js:187]: I suspect we want to hide all messages while in form template mode? If so, maybe a CSS solution is best?

**Review by pmusaraj — APPROVED** [review]

File: app/assets/javascripts/discourse/app/components/composer-messages.js (hunk 4 above)
Comment (CvX) [inline app/assets/javascripts/discourse/app/components/composer-messages.js:187]: @keegangeorge if `form_template_ids` is undefined then this fails with:

> TypeError: Cannot read properties of undefined (reading 'length')

**Pull request merged**